proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2357 @@
+  v`` converges, and *at most once* if it diverges.
+
+Where "consuming at most once" is defined by induction, like
+"consuming exactly once", but every sub-consumption is optional.
+
+In the paper, we gave a simplified specification of a linear ``IO``
+monad (called ``IOL``) which ignored the issue of exception for the
+sake of simplicity. Can we, still, write a resource-safe ``RIO`` monad
+with linear types despite the added difficulty of exceptions? Yes, as
+this section will show.
+
+Concretely, how do we ensure that the sockets from the example API are
+always closed, even in presence of exceptions? This boils down to how
+the ``RIO`` monad is implemented. Below is a sketch of one possible
+implementation of ``RIO`` (see `here
+<https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs>`_
+for a detailed implementation).
+
+First, note that since Haskell programs are of type ``IO ()``, we need a
+way to run ``RIO`` in an ``IO`` computation, this is provided by the
+function
+
+::
+
+  runRIO :: RIO (Unrestricted a) -> IO a
+
+In order to achieve resource safety in presence of exception, ``runRIO``
+is tasked with releasing any live resource in case of exception.
+
+To implement this, ``RIO`` keeps a table of release actions, to be used
+in case of exceptions. Each resource implemented in the ``RIO``
+abstraction registers a release action in the release action table
+when they are acquired.
+
+If no exception occurs, then all resources have been released by the
+program. In case an exception occurs, the program jumps to the handler
+installed by ``runRIO``, which releases the leftover resources.
+
+An alternative strategy would be to add terminators on every resources
+acquired in ``RIO``. Release in the non-exceptional case would still
+be performed by the program, and the GC would be responsible for
+releasing resources in case of exception. The release in case of
+exception would be, however, less timely.
+
+Can ``RIO`` have a ``catch``?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is possible to catch exceptions inside of ``RIO``. But in order to
+ensure resource safety, the type cannot be linear:
+
+::
+
+  catchL :: Exception e
+         => RIO (Unrestricted a)
+	 -> (e -> RIO (Unrestricted a))
+	 -> RIO (Unrestricted a)
+
+That is: no linear resource previously allocated (in particular linear
+variables which are not ``RIO`` resources) can be referenced in the
+body or the handler, and no resource allocated in the body or handler
+can be returned. In effect, ``catchL`` delimits a new scope, in which
+linear resources are isolated. To implement ``catchL``, we simply give
+it its own release action table, so that in case of exceptions all the
+local resources are released by ``catchL``, as ``runRIO`` does, before
+the handler is called. The original release action table is then
+reinstated.
+
+With this implementation, it is clear that capturing linear resources
```

Comment:
I didn't understand this paragraph. Are you saying that the `catchL` as described above is unsafe?

Reply:
It is supposed to mean that the implementation described in the previous paragraph cannot be typed with linear arrows. And that the body and handler must return `Unrestricted` results, because a returned linear value could contain resources. And these would escape their release-table, and would leak in case of exception further down the line.

I guess this warrant a drastic rephrasing, doesn't it?